### Module 4 - Probability Exercises 

Most of the concepts we discussed in this module were theoretical. Examples were used to explain how to apply conditional probability concepts on dataset columns. The exercises in this notebook will be similar to what you have seen in lab and practice notebooks. 

Refer to your labs: [Bayes](../labs/Bayes.ipynb), [Conditional_Probability](../labs/Conditional_Probability.ipynb) and [Other_Distributions](../labs/Other_Distributions.ipynb).  

Practice notebooks: [Conditional_Probability](../practices/Conditional_Probability.ipynb) and [Bayes](../labs/Bayes.ipynb). 

Let's recall what is an independent and dependent event through below activity. Consider an event of rolling two dice. The sample space S={{1,1}, {1,2}, {1,3},....,{6,6}} contains 36 possible combinations on two dice.

<span style="color:#1871d6; font-weight:400">Activity 1.a: </span> What is the probability of getting a 2 on either dice?

In [1]:
# Your code for activity 1.a goes here...

install.packages('prob')

library(prob)
S <- rolldie(2, makespace = TRUE)

# Define the event A such that it has all possible combinations including a 2 {{1,2}, {2,2}, {2,3},...} 


also installing the dependencies ‘spatial’, ‘gss’, ‘stabledist’, ‘timeSeries’, ‘fBasics’, ‘fOptions’, ‘combinat’, ‘fAsianOptions’

Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
Loading required package: combinat

Attaching package: ‘combinat’

The following object is masked from ‘package:utils’:

    combn

Loading required package: fAsianOptions
Loading required package: timeDate
Loading required package: timeSeries
Loading required package: fBasics
Loading required package: fOptions

Attaching package: ‘prob’

The following objects are masked from ‘package:base’:

    intersect, setdiff, union



In [20]:
head(S)

X1 X2 probs     
1 1  1  0.02777778
2 2  1  0.02777778
3 3  1  0.02777778
4 4  1  0.02777778
5 5  1  0.02777778
6 6  1  0.02777778

In [21]:
A <- subset(S, X1 == 2 | X2 == 2)

In [22]:
A

X1 X2 probs     
2  2  1  0.02777778
7  1  2  0.02777778
8  2  2  0.02777778
9  3  2  0.02777778
10 4  2  0.02777778
11 5  2  0.02777778
12 6  2  0.02777778
14 2  3  0.02777778
20 2  4  0.02777778
26 2  5  0.02777778
32 2  6  0.02777778

<span style="color:#1871d6; font-weight:400">Activity 1.b: </span> What is the probability of getting a 2 on either dice given that the sum of the two outcomes is less than 6? 

In [25]:
# Your code for activity 1.b goes here...

# Define the event A such that it has all possible combinations including 2 {{1,2}, {2,2},...}. 
A

# Define event B such that sum of outcomes is less than 6. 
B <- subset(S, (X1 + X2) < 6)
B
# Event A is conditional on Event B 

paste('P(A/B): ', Prob(A, given=B))

X1 X2 probs     
2  2  1  0.02777778
7  1  2  0.02777778
8  2  2  0.02777778
9  3  2  0.02777778
10 4  2  0.02777778
11 5  2  0.02777778
12 6  2  0.02777778
14 2  3  0.02777778
20 2  4  0.02777778
26 2  5  0.02777778
32 2  6  0.02777778

X1 X2 probs     
1  1  1  0.02777778
2  2  1  0.02777778
3  3  1  0.02777778
4  4  1  0.02777778
7  1  2  0.02777778
8  2  2  0.02777778
9  3  2  0.02777778
13 1  3  0.02777778
14 2  3  0.02777778
19 1  4  0.02777778

[1] "P(A/B):  0.5"

<span style="color:#1871d6; font-weight:400">Activity 2: </span> Toss a coin twice. The sample space is given by S = {HH, HT, TH, TT}. Let A = {head occurs} and B = {a head and tail occur}. What are the probabilities P(A|B) and P(B|A)?

In [26]:
HT <- iidspace(c("H","T"), ntrials = 2, probs = c(0.5, 0.5))
A <- subset(HT, X1 == 'H' | X2 == 'H')
B <- subset(HT, (X1=='H' & X2=='T') | (X1=='T' & X2=='H'))

In [27]:
paste('P(A|B): ', Prob(A, given=B))
paste('P(B|A): ', Prob(B, given=A))

[1] "P(A|B):  1"

[1] "P(B|A):  0.666666666666667"

````
P(A|B) = 1. This would be true as there has to be a heads in B.
P(B|A) = 0.67. This would be true as a H and T occur 2/3 of the time from A. The other 1/3 being HH in A.

````

<span style="color:#1871d6; font-weight:400">Activity 3: </span> We have data about the smoking status versus the gender of people working in a company.

<table>
    <tr>
        <th></th>
        <th colspan="2">gender</th>
        <th>sum</th>
    </tr>
    <tr>
        <td rowspan="2">smoke</td>
        <td>80</td>
        <td>54</td>
        <td>134</td>
    </tr>
    <tr>
        <td>15</td>
        <td>19</td>
        <td>34</td>
    </tr>
    <tr>
        <td>sum</td>
        <td>95</td>
        <td>73</td>
        <td>168</td>
    </tr>
</table>
    
    
3.a) If one person were selected at random from the data set, what is the probability that selected person is female?

3.b) What is the probability that selected person is a smoker?

In [47]:
# Code for Activity 3 goes here -----
# Generate the matrix required to find probabilities
S <- matrix(c(80, 15, 54, 19), 2)
rownames(S) <- c('nonsmoker', 'smoker')
colnames(S) <- c('female', 'male')
S

female male
nonsmoker 80     54  
smoker    15     19

In [48]:
addmargins(S)

female male Sum
nonsmoker 80     54   134
smoker    15     19    34
Sum       95     73   168

In [49]:
addmargins(prop.table(S))

female     male      Sum     
nonsmoker 0.47619048 0.3214286 0.797619
smoker    0.08928571 0.1130952 0.202381
Sum       0.56547619 0.4345238 1.000000

In [51]:
# Answer for Activity 3.a goes here -----
# P(Female) = 0.57
0.57

[1] 0.57

In [50]:
# Answer for Activity 3.b goes here -----
# P(smoker | Female) = P(smoker & female)/P(female)

15 / (80+15)

[1] 0.1578947

<span style="color:#1871d6; font-weight:400">Activity 4: </span>  Load the framingham data from the directory '/datasets/framingham'. Find out the probability of a randomly subject to have the risk of coronary heart disease given that the subject is a male.

In [52]:
# Code for Activity 4 goes here -----
framingham_data <- read.csv("/dsa/data/all_datasets/framingham/framingham.csv")
head(framingham_data)

male age education currentSmoker cigsPerDay BPMeds prevalentStroke
1 1    39  4         0              0         0      0              
2 0    46  2         0              0         0      0              
3 1    48  1         1             20         0      0              
4 0    61  3         1             30         0      0              
5 0    46  3         1             23         0      0              
6 0    43  2         0              0         0      0              
  prevalentHyp diabetes totChol sysBP diaBP BMI   heartRate glucose TenYearCHD
1 0            0        195     106.0  70   26.97 80         77     0         
2 0            0        250     121.0  81   28.73 95         76     0         
3 0            0        245     127.5  80   25.34 75         70     0         
4 1            0        225     150.0  95   28.58 65        103     1         
5 0            0        285     130.0  84   23.10 85         85     0         
6 1            0        228     180.0 110   30.30 77         99     0

In [81]:
# Creating the table
male_cor <- with(framingham_data, table(framingham_data$TenYearCHD, framingham_data$male))
colnames(male_cor) <- c('female', 'male')
rownames(male_cor) <- c('CHD -', 'CHD +')
male_cor.table <- as.table(male_cor)

In [82]:
addmargins(male_cor.table)
addmargins(prop.table(male_cor.table))

female male Sum 
  CHD - 2119   1477 3596
  CHD +  301    343  644
  Sum   2420   1820 4240

female     male       Sum      
  CHD - 0.49976415 0.34834906 0.8481132
  CHD + 0.07099057 0.08089623 0.1518868
  Sum   0.57075472 0.42924528 1.0000000

Using Bayes

CHD = CHD +

m = male

                            p(CHD) * p(m|CHD)
     p(CHD|m) =  -------------------------------------
               [p(CHD) * p(m|CHD)] + [ p(CHD') * p(m|CHD')]

In [85]:
#P(CHD|male) = P(CHD and male)/P(male) <- updated per professor's notes
# Enter your answer for Activity 4 -----
# using Bayes see above

(0.15 * 0.08) / ((0.15 * 0.08) + (0.85 * 0.35))

[1] 0.03877221

<span style="color:#1871d6; font-weight:400">Activity 5: </span> Find out the probability of randomly selected subject to have the risk of coronary heart disease given subject is less than or equal to 40 years of age.

In [78]:
#using these cells to make sure all the data in the tables is in the right columns/rows
#less_40 <- subset(framingham_data, framingham_data$age <=40)
#nrow(less_40)

[1] 748

In [90]:
#has_chd <- subset(framingham_data, framingham_data$TenYearCHD == 1)
#nrow(has_chd)

[1] 644

In [88]:
# Code for activity 5 goes here...
age_cor <- with(framingham_data, table(framingham_data$TenYearCHD, framingham_data$age <= 40))
colnames(age_cor) <- c('>40', '<=40')
rownames(age_cor) <- c('CHD -', 'CHD +')
age_cor <- as.table(age_cor)

In [91]:
addmargins(age_cor)
addmargins(prop.table(age_cor))

>40  <=40 Sum 
  CHD - 2886 710  3596
  CHD +  606  38   644
  Sum   3492 748  4240

>40       <=40        Sum      
  CHD - 0.6806604 0.167452830 0.8481132
  CHD + 0.1429245 0.008962264 0.1518868
  Sum   0.8235849 0.176415094 1.0000000

Using Bayes

CHD = CHD +

<=40 = less than or equal to 40

                            p(CHD) * p(<=40|CHD)
     p(CHD|m) =  -------------------------------------
               [p(CHD) * p(<=40|CHD)] + [ p(CHD') * p(<=40|CHD')]

In [93]:
# P(CHD | age <= 40) = P(CHD and age <= 40) / P(age <= 40)
# Enter your answer for Activity 5 -----

(0.15 * 0.01)/((0.15 * 0.01) + (0.85 * 0.17))


[1] 0.01027397

<span style="color:#1871d6; font-weight:400">Activity 6: </span> Find out the probability of randomly selected subject to have the risk of coronary heart disease given that the subject smokes less than 10 cigarettes and is a current smoker. 

**Hint:** Use a 3-way table to include the variables cigsPerDay, TenYearCHD and currentSmoker. Use currentSmoker as the 3rd dimension/input.

Per email, we are not considering current smokers and therefore do not need a 3-way table on this

In [96]:
#pack_10 <- subset(framingham_data, framingham_data$cigsPerDay < 10)
#nrow(pack_10)

[1] 2631

In [94]:
# Code for activity 6 goes here....
less_cigs <- with(framingham_data, table(framingham_data$TenYearCHD, framingham_data$cigsPerDay < 10))
colnames(less_cigs) <- c('>=10', '<10')
rownames(less_cigs) <- c('CHD -', 'CHD +')
less_cigs <- as.table(less_cigs)

In [95]:
addmargins(less_cigs)
addmargins(prop.table(less_cigs))

>=10 <10  Sum 
  CHD - 1302 2267 3569
  CHD +  278  364  642
  Sum   1580 2631 4211

>=10       <10        Sum      
  CHD - 0.30919022 0.53835194 0.8475422
  CHD + 0.06601757 0.08644028 0.1524578
  Sum   0.37520779 0.62479221 1.0000000

Using Bayes

CHD = CHD +

<10 = less than 10 packs a day

                            p(CHD) * p(<10|CHD)
     p(CHD|m) =  -------------------------------------
               [p(CHD) * p(<10|CHD)] + [ p(CHD') * p(<10|CHD')]

In [97]:
# P( risk of CHD | subject smokes <10 cigs ) = P( risk of CHD and subject smokes <10 cigs)/ P(subject smokes <10 cigs)
# Enter your answer for Activity 6 -----

(0.15*0.09)/((0.15*0.09)+(0.85 * 0.54))


[1] 0.02857143

<span style="color:#1871d6; font-weight:400">Activity 7.a: </span> Find out the probability of randomly selected subject to have the risk of coronary heart disease given that the subject has totChol > 300 and has BMI >30. 

<span style="color:#1871d6; font-weight:400">Activity 7.b: </span> Find out the probability of randomly selected subject to have the risk of coronary heart disease given that the subject has totChol > 300 and has BMI < 30. 

**Hint:** Use a 3-way table. Use BMI as the 3rd dimension/input.

In [99]:
?xtabs

In [100]:
# Code for activity 7 goes here....

# with(framingham_data,xtabs(~(totChol>300)+TenYearCHD+(BMI>30)))

CHD_BMI_totChol <- with(framingham_data,xtabs(~(totChol>300)+TenYearCHD+(BMI>30)))

In [103]:
addmargins(CHD_BMI_totChol)

, , BMI > 30 = FALSE

             TenYearCHD
totChol > 300    0    1  Sum
        FALSE 2909  465 3374
        TRUE   214   58  272
        Sum   3123  523 3646

, , BMI > 30 = TRUE

             TenYearCHD
totChol > 300    0    1  Sum
        FALSE  386   91  477
        TRUE    38   11   49
        Sum    424  102  526

, , BMI > 30 = Sum

             TenYearCHD
totChol > 300    0    1  Sum
        FALSE 3295  556 3851
        TRUE   252   69  321
        Sum   3547  625 4172


In [104]:
# Enter your answer for Activity 7.a (BMI > 30) -----  
# P( risk of CHD | subject has totChol > 300 ) = P( risk of CHD and subject has totChol > 300 )/ P(subject has totChol > 300 )

(102 * 11)/((102*11)+(424*38))

[1] 0.06510386

In [105]:
# Enter your answer for Activity 7.b (BMI < 30) -----  
# P( risk of CHD | subject has totChol > 300 ) = P( risk of CHD and subject has totChol > 300 )/ P(subject has totChol > 300 )
(523*58)/((523*58)+(3123*214))

[1] 0.04341765

<span style="color:#1871d6; font-weight:400">Activity 8: </span> Consider two machines producing memory chips. Machine A is in production 18 hours a day producing 80% of the products while machine B is in production for 6 hours a day producing 20% of the products. Both machines have a fault rate of 0.002 and 0.007, respectively. If a randomly selected chip is faulty, which machine could be the most probable source of producing that chip?

In [108]:
# Code for activity 8 goes here. 

# We don't have any prior probabilities for the machines. We consider amount of production they do as our prior.
prior_probs <- c(0.8, 0.2)

# Both machines A and B have a fault rate of 0.002 and 0.007. WE use this info as likelihood of generating a faulty chip
like <- c(0.002, 0.007)

# Generate posterior probabilities based on prior probability and likelihood of each event.
post <- prior_probs * like
post

[1] 0.0016 0.0014

In [109]:
post / sum(post)

[1] 0.5333333 0.4666667

Based on the above math, there is a 53.3% chance Machine A produced the faulty chip, and a 46.7% chance the faulty chip was produced by Machine B.

Therefore, it is more likely that the faulty chip was produced by Machine A.